File: src/linkage_sim/forces/assembly.py
"""Generalized force vector Q assembly.

Iterates all force elements attached to a mechanism and sums their
contributions into the global Q vector.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from linkage_sim.core.state import State
from linkage_sim.forces.protocol import ForceElement


def assemble_Q(
    state: State,
    force_elements: list[ForceElement],
    q: NDArray[np.float64],
    q_dot: NDArray[np.float64],
    t: float,
) -> NDArray[np.float64]:
    """Assemble the generalized force vector from all force elements.

    Args:
        state: Mechanism state (coordinate bookkeeping).
        force_elements: List of force elements to evaluate.
        q: Generalized coordinate vector (n_coords,).
        q_dot: Generalized velocity vector (n_coords,).
        t: Current time.

    Returns:
        Total generalized force vector Q (n_coords,).
    """
    Q = np.zeros(state.n_coords)

    for fe in force_elements:
        Q += fe.evaluate(state, q, q_dot, t)

    return Q

File: linkage-sim-rs/src/forces/assembly.rs
//! Assemble total generalized force Q from all force elements.

use nalgebra::DVector;

use crate::core::state::State;
use crate::forces::gravity::Gravity;

/// Assemble total generalized force Q from gravity (and future force elements).
///
/// Currently supports Gravity. Additional force element types
/// will be added as the port progresses (springs, dampers, etc.).
pub fn assemble_q(
    state: &State,
    gravity: Option<&Gravity>,
    q: &DVector<f64>,
    q_dot: &DVector<f64>,
    t: f64,
) -> DVector<f64> {
    let mut total = DVector::zeros(state.n_coords());

    if let Some(g) = gravity {
        total += g.evaluate(state, q, q_dot, t);
    }

    total
}
